File: solution/ts/week05-complexity/exercise2/src/commandProcessor.ts
import {Command} from "./command";

export class CommandProcessor {
    private commands: Map<string, Command> = new Map<string, Command>();

    get commandsNames(): Set<string> {
        return new Set(this.commands.keys());
    }

    addCommand(commandName: string, command: Command): void {
        this.commands.set(commandName, command);
    }

    processCommand(commandName: string, printTo: (message: string) => void): void {
        const cmd = this.commands.get(commandName) || { executeAndDisplayResult: () => "Command not recognized." };
        printTo(cmd.executeAndDisplayResult());
    }
}

File: solution/c#/week05-complexity/CommandProcessor/CommandProcessor.cs
namespace CommandProcessor
{
    public class CommandProcessor
    {
        private readonly Dictionary<string, ICommand> _commands = [];
        public IEnumerable<string> CommandsNames() => _commands.Keys;

        public void AddCommand(string commandName, ICommand command)
            => _commands[commandName] = command;

        public void ProcessCommand(string commandName, Action<string> printTo) =>
            printTo(_commands.TryGetValue(commandName, out var cmd)
                ? cmd.ExecuteAndDisplayResult()
                : "Command not recognized.");
    }
}